libvast/vast/transform_steps/projection.hpp

```diff
@@ -0,0 +1,38 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#include "vast/fwd.hpp"
+
+#include "vast/transform.hpp"
+
+#include <unordered_set>
+namespace vast {
+
+/// Projects the input onto the specified fields(deletes unspecified fields).
+class projection_step : public generic_transform_step,
+                        public arrow_transform_step {
+public:
+  projection_step(std::vector<std::string> fields);
+
+  caf::expected<table_slice> operator()(table_slice&& slice) const override;
+
+  caf::expected<std::pair<type, std::shared_ptr<arrow::RecordBatch>>>
+  operator()(type layout,
+             std::shared_ptr<arrow::RecordBatch> batch) const override;
+
+private:
+  /// adjust the layout according to the projection
+  [[nodiscard]] caf::expected<std::pair<vast::type, std::vector<int>>>
+  adjust_layout(const vast::type& layout) const;
+
+  const std::vector<std::string> fields_;
+};
```

Comment:
Please add doxygen-esque comments above all functions and member variables. This is especially important for the `adjust_layout` function, whose return type isn't intuitive from the function name.